### PR TITLE
OX Quiz and more triggers

### PR DESCRIPTION
**Pull request by Zintixx**
This PR basically introduces and fixes some trigger functions.
## OX Quiz

 - OX Quiz triggers should now work.
 - Parsing over 1300 questions.
	 - Normally, these questions are categorized to different tiers (for different rounds) but it's hard to know how these were categorized. For now, all questions have an equal chance to be picked.

## Triggers

 - Parsing more trigger entities from the data (Ropes, Ladders, Cubes, Sounds)
 - Parsing PatrolData and Waypoints. These are used for triggers to move NPCs/Mobs
	 - Currently these are not implemented.

## Etc

 - Added `RewardContentsParser` to be able to find what rewards are given for certain contents (ex. Kay's Coin for Global Events
 - Added `EventCommands`
	 - A command to allow for the global event portals (aka popup) to pop up.
	 - To force start a Global Event, type `/event` and it'll ask you to pick 3 events that you wish for it to pop up globally for all players.
- Fixed `GetSessions` in Party. Parties should now correctly disband upon all members logging out.

### CODE REVIEW

File: GameDataParser/Program.cs [hunk 1]
```diff
@@ -65,7 +65,9 @@ private static async Task Main()
                 new FurnishingShopParser(resources),
                 new HomeTemplateParser(resources),
                 new MasteryUGCHousingParser(resources),
-                new JobParser(resources)
+                new JobParser(resources),
+                new RewardContentParser(resources),
+                new OXQuizParser()
```
Comment (AlanMorel) [inline GameDataParser/Program.cs:70]: Does it not require resource?

File: MapleServer2/Commands/Game/EventCommands.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            AddParameter<string[]>("id", "Event Id");
+        }
+
+        public override void Execute(GameCommandTrigger trigger)
+        {
+            string[] args = trigger.Get<string[]>("id");
+
+            if (args == null || args.Length <= 1)
+            {
+                trigger.Session.SendNotice("No events provided. Choose 3 events. 1 = OX Quiz \n" +
```
Comment (AlanMorel) [inline MapleServer2/Commands/Game/EventCommands.cs:25]: Can we just have a 13 element string array to nicely display this? Could include just the text, like `Trap Master`, then use the index to count it, and then join the array using `\n`

File: MapleServer2/Triggers/UITriggers.cs [hunk 3]
```diff
@@ -14,24 +14,70 @@ public partial class TriggerContext
...
+                            // TODO: Unknown
+                            break;
+                        case "PickQuiz":
+                            // TODO: Use args to find a tier of a question
+                            OXQuizMetadata question = OXQuizMetadataStorage.GetQuestion();
+                            widget.OXQuizQuestion = new OXQuizQuestion(question.Category, question.QuestionText, question.AnswerText, question.Answer);
+                            if (question.Answer)
+                            {
+                                Field.BroadcastPacket(NoticePacket.Notice("Answer is true", NoticeType.ChatAndFastText));
+
```
Comment (AlanMorel) [inline MapleServer2/Triggers/UITriggers.cs:50]: Extra line, same on 55

File: MapleServer2/Types/TriggerObject.cs [hunk 4]
```diff
@@ -47,4 +49,48 @@ public TriggerActor(int id, bool isVisible, string stateName) : base(id)
             StateName = stateName;
         }
     }
+
+    public class TriggerLadder : TriggerObject
+    {
+        public bool IsVisible;
```
Comment (AlanMorel) [inline MapleServer2/Types/TriggerObject.cs:55]: Don't need `IsVisible` if it's already in the parent object, same on line 68 and 81

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: GameDataParser/Program.cs (hunk 1 above)
Comment (Zintixx) [reply GameDataParser/Program.cs:70]: No, it's not from the game data. This is all server-sided data that I grabbed from a ms2 wiki

File: MapleServer2/Types/TriggerObject.cs (hunk 4 above)
Comment (Zintixx) [reply MapleServer2/Types/TriggerObject.cs:55]: The parent object doesn't have `IsVisible`

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 5]
```diff
@@ -146,5 +168,25 @@ public static List<MapTriggerActor> GetTriggerActors(int mapId)
...
+
+        public static List<MapTriggerRope> GetTriggerRopes(int mapId)
+        {
+            return TriggerRopes.GetValueOrDefault(mapId);
+        }
+
+        public static List<MapTriggerSound> GetTriggerSounds(int mapId)
+        {
+            return TriggerSounds.GetValueOrDefault(mapId);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:190]:
```suggestion
        public static List<MapTriggerSound> GetTriggerSounds(int mapId) => TriggerSounds.GetValueOrDefault(mapId);
```

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 6]
```diff
@@ -146,5 +168,25 @@ public static List<MapTriggerActor> GetTriggerActors(int mapId)
...
+
+        public static List<MapTriggerLadder> GetTriggerLadders(int mapId)
+        {
+            return TriggerLadders.GetValueOrDefault(mapId);
+        }
+
+        public static List<MapTriggerRope> GetTriggerRopes(int mapId)
+        {
+            return TriggerRopes.GetValueOrDefault(mapId);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:185]:
```suggestion
        public static List<MapTriggerRope> GetTriggerRopes(int mapId) => TriggerRopes.GetValueOrDefault(mapId);
```

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 7]
```diff
@@ -146,5 +168,25 @@ public static List<MapTriggerActor> GetTriggerActors(int mapId)
...
+
+        public static List<MapTriggerCube> GetTriggerCubes(int mapId)
+        {
+            return TriggerCubes.GetValueOrDefault(mapId);
+        }
+
+        public static List<MapTriggerLadder> GetTriggerLadders(int mapId)
+        {
+            return TriggerLadders.GetValueOrDefault(mapId);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:180]:
```suggestion
        public static List<MapTriggerLadder> GetTriggerLadders(int mapId) => TriggerLadders.GetValueOrDefault(mapId);
```

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 8]
```diff
@@ -146,5 +168,25 @@ public static List<MapTriggerActor> GetTriggerActors(int mapId)
         {
             return TriggerActors.GetValueOrDefault(mapId);
         }
+
+        public static List<MapTriggerCube> GetTriggerCubes(int mapId)
+        {
+            return TriggerCubes.GetValueOrDefault(mapId);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:175]:
```suggestion
        public static List<MapTriggerCube> GetTriggerCubes(int mapId) => TriggerCubes.GetValueOrDefault(mapId);
```

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 9]
```diff
@@ -115,6 +127,16 @@ public static List<CoordS> GetHealingSpot(int mapId)
             return healthSpot.GetValueOrDefault(mapId);
         }
 
+        public static PatrolData GetPatrolData(int mapId, string patrolDataName)
+        {
+            return PatrolDatas.GetValueOrDefault(mapId).FirstOrDefault(x => x.Name == patrolDataName);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:133]: Now that you modified this class, you can change the method with one single return and use expression body.
```suggestion
        public static PatrolData GetPatrolData(int mapId, string patrolDataName) => PatrolDatas.GetValueOrDefault(mapId).FirstOrDefault(x => x.Name == patrolDataName);
```

File: MapleServer2/Data/Static/MapEntityStorage.cs [hunk 10]
```diff
@@ -115,6 +127,16 @@ public static List<CoordS> GetHealingSpot(int mapId)
...
+
+        public static WayPoint GetWayPoint(int mapId, string id)
+        {
+            return WayPoints.GetValueOrDefault(mapId).FirstOrDefault(x => x.Id == id);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Data/Static/MapEntityStorage.cs:138]:
```suggestion
        public static WayPoint GetWayPoint(int mapId, string id) => WayPoints.GetValueOrDefault(mapId).FirstOrDefault(x => x.Id == id);
```

File: MapleServer2/Tools/GlobalEventManager.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+        public GlobalEventManager()
+        {
+            GlobalEventList = new Dictionary<int, GlobalEvent>();
+        }
+
+        public void AddEvent(GlobalEvent globalEvent)
+        {
+            GlobalEventList.Add(globalEvent.Id, globalEvent);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Tools/GlobalEventManager.cs:17]:
```suggestion
        public void AddEvent(GlobalEvent globalEvent) => GlobalEventList.Add(globalEvent.Id, globalEvent);
```

File: MapleServer2/Tools/GlobalEventManager.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+        public void RemoveEvent(GlobalEvent globalEvent)
+        {
+            GlobalEventList.Remove(globalEvent.Id);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Tools/GlobalEventManager.cs:22]:
```suggestion
        public void RemoveEvent(GlobalEvent globalEvent) => GlobalEventList.Remove(globalEvent.Id);
```

File: MapleServer2/Tools/GlobalEventManager.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+        public int GetCount()
+        {
+            return GlobalEventList.Count;
+        }
```
Comment (Sparkymod) [inline MapleServer2/Tools/GlobalEventManager.cs:27]:
```suggestion
        public int GetCount() => GlobalEventList.Count;
```

File: MapleServer2/Tools/GlobalEventManager.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+        public GlobalEvent GetEventById(int id)
+        {
+            return GlobalEventList.GetValueOrDefault(id);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Tools/GlobalEventManager.cs:32]:
```suggestion
        public GlobalEvent GetEventById(int id) => GlobalEventList.GetValueOrDefault(id);
```

File: MapleServer2/Servers/Game/FieldManager.cs [hunk 15]
```diff
@@ -711,12 +750,34 @@ public int Decrement()
...
+
+        public Widget GetWidget(WidgetType type)
+        {
+            return Widgets.FirstOrDefault(x => x.Type == type);
+        }
+
+        public void EnableSceneSkip(bool enable)
+        {
+            SkipScene = enable;
+        }
```
Comment (Sparkymod) [inline MapleServer2/Servers/Game/FieldManager.cs:781]:
```suggestion
        public void EnableSceneSkip(bool enable) => SkipScene = enable;
```

File: MapleServer2/Servers/Game/FieldManager.cs [hunk 16]
```diff
@@ -711,12 +750,34 @@ public int Decrement()
...
+
+        public void AddWidget(Widget widget)
+        {
+            Widgets.Add(widget);
+        }
+
+        public Widget GetWidget(WidgetType type)
+        {
+            return Widgets.FirstOrDefault(x => x.Type == type);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Servers/Game/FieldManager.cs:776]:
```suggestion
        public Widget GetWidget(WidgetType type) => Widgets.FirstOrDefault(x => x.Type == type);
```

File: MapleServer2/Servers/Game/FieldManager.cs [hunk 17]
```diff
@@ -711,12 +750,34 @@ public int Decrement()
...
 
         public MapTimer GetMapTimer(string id)
         {
             return MapTimers.FirstOrDefault(x => x.Id == id);
         }
+
+        public void AddWidget(Widget widget)
+        {
+            Widgets.Add(widget);
+        }
```
Comment (Sparkymod) [inline MapleServer2/Servers/Game/FieldManager.cs:771]:
```suggestion
        public void AddWidget(Widget widget) => Widgets.Add(widget);
```

File: MapleServer2/Servers/Game/FieldManager.cs [hunk 18]
```diff
@@ -711,12 +750,34 @@ public int Decrement()
...
+                return;
+            }
+
             MapTimers.Add(timer);
         }
 
         public MapTimer GetMapTimer(string id)
         {
             return MapTimers.FirstOrDefault(x => x.Id == id);
         }
```
Comment (Sparkymod) [inline MapleServer2/Servers/Game/FieldManager.cs:766]:
```suggestion
        public MapTimer GetMapTimer(string id) => MapTimers.FirstOrDefault(x => x.Id == id);
```

**Review by Sparkymod — CHANGES_REQUESTED** [review]
Overall good, try to do expression body whenever you can.

File: MapleServer2/Triggers/MiniGameTriggers.cs [hunk 19]
```diff
@@ -10,18 +16,60 @@ public void StartMiniGame(int boxId, byte round, MiniGame type, bool isShowResul
...
+            List<IFieldObject<Player>> boxPlayers = new List<IFieldObject<Player>>();
+            foreach (IFieldObject<Player> player in Field.State.Players.Values)
+            {
+                if (FieldManager.IsPlayerInBox(box, player))
+                {
+                    boxPlayers.Add(player);
+                }
+            }
+
+            Random random = new Random();
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/MiniGameTriggers.cs:69]: Use **RandomProvider** class

File: MapleServer2/Triggers/PlayerTriggers.cs [hunk 20]
```diff
@@ -47,18 +50,51 @@ public void MoveToPortal(int userTagId, int portalId, int boxId)
...
+                return;
+            }
+
+            MapPortal dstPortal = MapEntityStorage.GetPortals(mapId).FirstOrDefault(portal => portal.Id == triggerId);
             foreach (IFieldObject<Player> player in players)
             {
-                //       player.Value.Warp(mapId, portal.Coord, portal.Rotation);
+                player.Value.Warp(mapId, dstPortal.Coord.ToFloat(), dstPortal.Rotation.ToFloat());
             }
+
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/PlayerTriggers.cs:97]: blank line

File: MapleServer2/Triggers/PlayerTriggers.cs [hunk 21]
```diff
@@ -93,12 +130,61 @@ public void SetPcRotation(Vector3 rotation)
...
+                            player.Value.TrophyUpdate(23400002, 1);
+                            break;
+                        case "finalsurvivor_start":
+                            player.Value.TrophyUpdate(23400003, 1);
+                            break;
+                        case "finalsurvivor_win":
+                            player.Value.TrophyUpdate(23400004, 1);
+                            break;
+                        default:
+                            Console.WriteLine($"Unhandled trophy set: {trophySet}");
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/PlayerTriggers.cs:178]: I think we should use Logger instead.

File: MapleServer2/Triggers/TriggerContext.cs [hunk 22]
```diff
@@ -116,7 +117,13 @@ public int GetUserCount(int boxId, int userTagId)
...
         public int GetUserValue(string key)
         {
-            return 0;
+            IFieldObject<Player> player = Field.State.Players.Values.FirstOrDefault(x => x.Value.Triggers.Any(x => x.Key == key));
+            if (player == null)
+            {
+                return 0;
+            }
+            int value = player.Value.Triggers.FirstOrDefault(x => x.Key == key).Value;
+            return value;
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/TriggerContext.cs:126]:
```suggestion
            return player.Value.Triggers.FirstOrDefault(x => x.Key == key).Value;
```

File: MapleServer2/Triggers/UITriggers.cs [hunk 23]
```diff
@@ -14,24 +14,60 @@ public partial class TriggerContext
...
-                player.Value.Widgets.Add(widget);
-            }
+            Field.AddWidget(widget);
         }
 
         public void WidgetAction(WidgetType type, string name, string args, int widgetArgNum)
         {
-            List<IFieldObject<Player>> players = Field.State.Players.Values.ToList();
-            foreach (IFieldObject<Player> player in players)
+            Console.WriteLine($"Doing Widget action on Widget Type: {type}. Action Name: {name}");
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/UITriggers.cs:22]: debug message

File: MapleServer2/Triggers/UITriggers.cs [hunk 24]
```diff
@@ -87,61 +124,68 @@ public void ScoreBoardSetScore(bool score)
...
-            }
-
-            if (box == "0")
-            {
-                Field.BroadcastPacket(MassiveEventPacket.TextBanner(type, script, duration));
-                return;
-            }
-
-            if (box.Contains('!'))
+            if (typeId != 0)
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/UITriggers.cs:127]: Can you invert please?

File: MapleServer2/Types/TriggerObject.cs (hunk 4 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/TriggerObject.cs:55]: I think he meant to add `IsVisible` to **TriggerObject** ?

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/TriggerObject.cs (hunk 4 above)
Comment (Zintixx) [reply MapleServer2/Types/TriggerObject.cs:55]: Not all TriggerObjects use `IsVisible`

Comment (AngeloTadeucci) [reply MapleServer2/Types/TriggerObject.cs:55]: My comment is gone, lol.. but sounds good!

File: MapleServer2/Triggers/PlayerTriggers.cs (hunk 21 above)
Comment (Sparkymod) [reply MapleServer2/Triggers/PlayerTriggers.cs:178]: Agree with this suggestion.

File: MapleServer2/Types/TriggerObject.cs (hunk 4 above)
Comment (AlanMorel) [reply MapleServer2/Types/TriggerObject.cs:55]: Oh, nevermind

File: MapleServer2/Triggers/UITriggers.cs [hunk 25]
```diff
@@ -124,68 +123,67 @@ public void ScoreBoardSetScore(bool score)
...
-                }
-
-                if (box == "0")
-                {
-                    Field.BroadcastPacket(MassiveEventPacket.TextBanner(type, script, duration));
+                    Field.BroadcastPacket(MassiveEventPacket.RoundBar(int.Parse(ids[0]), int.Parse(ids[1]), 1));
                     return;
                 }
+                Field.BroadcastPacket(MassiveEventPacket.RoundBar(int.Parse(ids[0]), int.Parse(ids[1]), int.Parse(ids[2])));
+            }
```
Comment (AngeloTadeucci) [inline MapleServer2/Triggers/UITriggers.cs:136]: missing `return` ?

**Review by AngeloTadeucci — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

**Pull request merged**